src/include/homestore/blk.h

```diff
@@ -24,122 +24,162 @@
-static_assert(sizeof(chunk_num_t) == (CHUNK_NUM_BITS - 1) / 8 + 1, "Expected blk_count_t to matching CHUNK_NUM_BITS");
-
-typedef uint8_t blk_temp_t;
-
-/* This structure represents the application wide unique block number. It also encomposses the number of blks. */
+#pragma pack(1)
 struct BlkId {
-private:
-    static constexpr uint64_t s_blk_num_mask{(static_cast< uint64_t >(1) << BLK_NUM_BITS) - 1};
-    static constexpr uint64_t s_nblks_mask{(static_cast< uint64_t >(1) << NBLKS_BITS) - 1};
-    static constexpr uint64_t s_chunk_num_mask{(static_cast< uint64_t >(1) << CHUNK_NUM_BITS) - 1};
+protected:
+    struct serialized {
+        blk_num_t m_is_multi : 1; // Is it a part of multi blkid or not
+        blk_num_t m_blk_num : 31; // Block number which is unique within the chunk
+        blk_count_t m_nblks;      // Number of blocks+1 for this blkid, don't directly acccess this - use blk_count()
+        chunk_num_t m_chunk_num;  // Chunk number - which is unique for the entire application
 
-public:
-    static constexpr blk_count_t max_blks_in_op() { return (1 << NBLKS_BITS); }
-    static constexpr uint64_t max_id_int() { return (1ull << (BLK_NUM_BITS + NBLKS_BITS + CHUNK_NUM_BITS)) - 1; }
+        serialized() : m_is_multi{0}, m_blk_num{0}, m_nblks{0}, m_chunk_num{0} {}
+        serialized(bool is_multi, blk_num_t blk_num, blk_count_t nblks, chunk_num_t cnum) :
+                m_is_multi{is_multi ? 0x1u : 0x0u}, m_blk_num{blk_num}, m_nblks{nblks}, m_chunk_num{cnum} {}
+    };
+    static_assert(sizeof(serialized) == sizeof(uint64_t), "Expected serialized size to 64 bits");
 
-    static int compare(const BlkId& one, const BlkId& two);
-    uint64_t to_integer() const;
+    serialized s;
 
+public:
+    BlkId() = default;
     explicit BlkId(uint64_t id_int);
-    BlkId(blk_num_t blk_num, blk_count_t nblks, chunk_num_t chunk_num = 0);
-    BlkId() { invalidate(); }
-    BlkId(const BlkId&) = default;
-    BlkId& operator=(const BlkId&) = default;
+    BlkId(blk_num_t blk_num, blk_count_t nblks, chunk_num_t chunk_num);
+    BlkId(BlkId const&) = default;
+    BlkId& operator=(BlkId const&) = default;
     BlkId(BlkId&&) noexcept = default;
     BlkId& operator=(BlkId&&) noexcept = default;
-    bool operator==(const BlkId& other) const { return (compare(*this, other) == 0); }
-    bool operator>(const BlkId& other) const { return (compare(*this, other) > 0); }
-    bool operator<(const BlkId& other) const { return (compare(*this, other) < 0); }
+
+    bool operator==(BlkId const& other) const { return (compare(*this, other) == 0); }
+    bool operator>(BlkId const& other) const { return (compare(*this, other) > 0); }
+    bool operator<(BlkId const& other) const { return (compare(*this, other) < 0); }
+
+    blk_num_t blk_num() const { return s.m_blk_num; }
+    blk_count_t blk_count() const { return s.m_nblks; }
+    chunk_num_t chunk_num() const { return s.m_chunk_num; }
+    bool is_multi() const { return s.m_is_multi; }
```

Comment:
can this be moved to MultiBlkId?

Reply:
MultiBlkId is not a vtable based dervied class, it is a non-virtual derived class. This is done for backward compatibility where upper layer refers them as BlkId and we need to see if they are multi or not (from BlkId reference). It is used to assert that and we would also need to serialize the BlkId. Hence this should stay in BlkId.